django_prices/models.py

```diff
@@ -1,91 +1,125 @@
+        currency_field="price_currency",
+        verbose_name=None,
+        **kwargs
+    ):
+        super(MoneyField, self).__init__()
+        self.amount_field = amount_field
+        self.currency_field = currency_field
+
+    def __str__(self):
+        return "MoneyField(amount_field=%s, currency_field=%s)" % (
+            self.amount_field,
+            self.currency_field,
+        )
+
+    def __get__(self, instance, cls=None):
+        if instance is None:
+            return self
+
+        amount = getattr(instance, self.amount_field)
+        currency = getattr(instance, self.currency_field)
+        if amount and currency:
+            return Money(amount, currency)
+        return self.get_default()
+
+    def __set__(self, instance, value):
+        amount = None
+        currency = None
         if value is not None:
-            return value
-        return super(MoneyField, self).value_to_string(value)
+            amount = value.amount
+            currency = value.currency
+        setattr(instance, self.amount_field, amount)
+        setattr(instance, self.currency_field, currency)
 
     def formfield(self, **kwargs):
-        defaults = {"currency": self.currency, "form_class": forms.MoneyField}
-        defaults.update(kwargs)
-        return super(MoneyField, self).formfield(**defaults)
+        available_currencies = []
+        if hasattr(self, "model"):
+            available_currencies = self.model._meta.get_field(
+                self.currency_field
+            ).choices
+        return forms.MoneyField(available_currencies=available_currencies)
 
     def get_default(self):
-        default = super(MoneyField, self).get_default()
-        return self.to_python(default)
+        default_currency = None
+        default_amount = Decimal(0)
+        if hasattr(self, "model"):
+            default_currency = self.model._meta.get_field(
+                self.currency_field
+            ).get_default()
+            default_amount = self.model._meta.get_field(self.amount_field).get_default()
 
-    @cached_property
-    def validators(self):
-        validators = list(itertools.chain(self.default_validators, self._validators))
-        return validators + [
-            MoneyPrecisionValidator(self.currency, self.max_digits, self.decimal_places)
-        ]
+        return Money(default_amount, default_currency)
 
-    def deconstruct(self):
-        name, path, args, kwargs = super(MoneyField, self).deconstruct()
-        kwargs["currency"] = self.currency
-        return name, path, args, kwargs
 
-
-class TaxedMoneyField(object):
+class TaxedMoneyField(NonDatabaseFieldBase):
 
     description = "A field that combines net and gross fields values into TaxedMoney."
-    empty_values = list(validators.EMPTY_VALUES)
-
-    # Field flags
-    auto_created = False
-    blank = True
-    concrete = False
-    editable = False
-
-    is_relation = False
-    remote_field = None
 
     def __init__(
```

Comment:
@korycins and I are wondering whether this should be changed to accept `net_amount_field`, `gross_amount_field` and `currency_field` instead of the current structure that forces you to create three virtual fields when you only need one.

Imagine the situation as described in the readme:

```python

class Model(models.Model):
    currency = models.CharField(max_length=3, default="BTC")
    price_net_amount = models.DecimalField(max_digits=9, decimal_places=2, default="5")
    price_net = MoneyField(amount_field="price_net_amount", currency_field="currency")
    price_gross_amount = models.DecimalField(
        max_digits=9, decimal_places=2, default="5"
    )
    price_gross = MoneyField(
        amount_field="price_gross_amount", currency_field="currency"
    )
    price = TaxedMoneyField(net_field='price_net', gross_field='price_gross')
```

What should be the result of the following code:

```python
x = Model()
x.price_net = Money(5, 'USD')
x.price_gross = Money(7, 'EUR')
print(x.price)
y = Model(price_net=Money(1, 'PLN'), price_gross=(2, 'EUR'))
print(y.price)
```

Reply:
I've refactored that.